gossip/src/duplicate_shred_handler.rs

```diff
@@ -90,17 +97,43 @@ impl DuplicateShredHandler {
     pub fn new(
         blockstore: Arc<Blockstore>,
         leader_schedule_cache: Arc<LeaderScheduleCache>,
+        bank_forks: Arc<RwLock<BankForks>>,
     ) -> Self {
         Self {
             chunk_map: HashMap::new(),
             validator_pending_proof_map: HashMap::new(),
             last_root: 0,
+            last_root_cleaned: 0,
+            cached_on_slot: 0,
+            cached_staked_nodes: None,
+            cached_slots_in_epoch: 100,
             blockstore,
             leader_schedule_cache,
+            bank_forks,
             cleanup_count: CLEANUP_EVERY_N_LOOPS,
         }
     }
 
+    fn cache_root_info(&mut self) {
+        let last_root = self.blockstore.last_root();
+        if self.cached_staked_nodes.is_none() || last_root != self.last_root {
+            self.last_root = last_root;
+            // The stuff we need only changes per epoch, so we only need to cache
+            // once per epoch, and it's okay if we use old data in a few slots.
+            if self.cached_staked_nodes.is_some()
+                && last_root < self.cached_slots_in_epoch + self.cached_on_slot
+            {
+                return;
+            }
```

Comment:
instead of this you can store the `epoch` when things were cached and just check if the epoch has changed.

Reply:
Can I get current epoch without reading the root bank? I am interested in checking whether epoch changed without reading the bank which involves a read lock, currently I'm using cached_on_slot not cached_on_epoch just because I'm getting the slot number for free without a read lock.

Reply:
Changed to RwLock::try_read now, it should be okay if we can't get data for new epoch for a few slots, we just use the old data.